README

This is alternative firmware for model radio FlySky FS-GT3B/C and clones.
This software comes with absolutely NO-WARRANTY.



ChangeLog
*0.6.2 ()
    added global option to select between 1 or 2 detents of rotate encoder,
	which must be applied to change value
    for 4WS/DIG/BRAKE, enable it directly after choosing channel, previously
	it was after leaving menu

*0.6.1 (18 May 2012)
    repaired ABS to use cycle length same as before PPM frame length changes
    added extra brake channel for throttle brake side
    repaired steering/throttle/channel speed to not beeing dependent
	on ppm frame length
    added buttons LOCK until ENTER-long to global menu
    repaired key setting, buttons long RS and EN were unaccesible

*0.6.0 (28 Apr 2012)
    increased model memories by using FLASH also
    added more Multi-Positions

*0.5.1 (26 Apr 2012)
    added possibility to connect potentiometer instead of CH3 button

*0.5.0 (23 Apr 2012)
    global menu was restructuralized
    added ability to select number of channels (2-MAX_CHANNELS) for each
	model, less channels means less delay
    poweron warn when not-centered now has separate global option
    added up/down/lap timers with throttle trigger start
    added BRK button function to override throttle and apply full brake
    big internal changes handling different menus, key mapping uses
	2-char option identification instead of various symbols
    added global option rotate_reverse to switch direction of rotate
	endoder at GT3C
    added globally settable PPM constant sync/constant frame + selection
	of length
    added new menu to directly set channel 3-8 value

*0.4.2 (13 Apr 2012)
    added global setting to disable center/reset value beep
    minimum battery voltage set to 2V for those experimenting with 1S cell
    PPM signal now uses constant length of SYNC signal to get it to
	receiver as soon as possible, also servo values are calculated
	just before generating servo PPM signals to have lowest
	steering/throttle to PPM signal delay
    ADC measuring every 1ms, averaging last 4 samples when used,
	globally settable to use last 4/1 values
    added poweron beep, can be set ON/OFF in global config
    at poweron, if steering or throttle are not in center dead zone,
	beep 3 times if poweron beep is allowed
    inactivity timer compares ADC values with previous ones, so now
	it will activate when radio is not center callibrated also
    added key function BLS to shutup battery low beeper, by default is
	assigned to END-long
    3-pos switch instead of CH3 button now works

*0.4.1 (17 Aug 2011)

    minimum battery voltage set to 6V
    CH3 key set default to OFF
    added DGR key function to reset DIG mix to centre
    added posibility to select DIG instead of channel number at MultiPosition
    added lap count functions assignable to keys

*0.4.0 (18 Jul 2011)

    use steering trim when dual-ESC steering
    added channel_speed for all channels, throttle channel can be applied only
	for forward side
    repaired setting to default values for functions of config-values mapped
	to trims, this should be done only for non-config values (channel,
	mix)

*0.3.2 (27 Jun 2011)

    added option throttle brake cut of to be able to drive forward only
	vehicles (boats)
    added MultiPosition settable to channel

*0.3.1 (26 Jun 2011)

    DIG mix can be set to channel 1 to have dual-ESC steering (boat, tank, ...)

*0.3.0 (24 Jun 2011)

    added globally settable key long-press delay
    when using trims, longer beep at value reset and short pause ignoring keys
    momentary trims/keys now sets value only when state of buttons changed
	(previously it always forced value to reflect button state), so it is
	possible to map momentary and non-momentary keys to the same
	channel/mix value, also value is showed when changed
    added possibility for momentary trims and 2-state keys to return not
	to centre/left, but to previous value which was active before button
	press

*0.2.0 (9 Jun 2011)

    added 4WS mix
    added DIG mix
    added steering speed (Turn, Return)
    added inactivity alarm
    added function C3R for keys to reset channel3 value to centre
    internal changes of trims momentary/opposite_reset

*0.1.0 (2 Jun 2011)

    added special chars "-1" and "10" to be able to display values -199...1099
    both trims-long now only resets, before there was 1 press of first button
    changed EEPROM init to eliminate broken initialization
    in menu: use ENTER to switch items and END/ENTER-long to end of edit of item
    added key mapping to configuration of model and removed coresponding
      global settings (trim_step, autorepeat, ch3_momentary)


*0.0.0 (25 May 2011)

    initial public revision



lcd.h
/*
    lcd - include file
    Copyright (C) 2011 Pavel Semerad

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/


#ifndef _LCD_INCLUDED
#define _LCD_INCLUDED


#include "gt3b.h"


// manipulating individual segments, pos is one of 128 segments
//   as special value (see below)
#define LS_OFF	0
#define LS_ON	1
extern void lcd_segment(u8 pos, u8 on_off);
#define LB_OFF	0
#define LB_SPC	1
#define LB_INV	2
extern void lcd_segment_blink(u8 pos, u8 on_off);

// arrays of segment positions for lcd_segment()
// for chars and 7seg is order from top left to down, then next column
// for menu is order from top left to right, then next line
#define LCD_CHAR_COLS 5
#define LCD_CHAR_ROWS 7
#define LCD_CHAR_BYTES (LCD_CHAR_COLS * LCD_CHAR_ROWS)
extern const u8 lcd_seg_char1[];
extern const u8 lcd_seg_char2[];
extern const u8 lcd_seg_char3[];
extern const u8 lcd_seg_7seg[];
extern const u8 lcd_seg_menu[];
// lcd segments addresses for symbols
#define LS_SYM_LOWPWR	0xdf
#define LS_SYM_CHANNEL	0xde
#define LS_SYM_MODELNO	0xdd
#define LS_SYM_PERCENT	0xc7
#define LS_SYM_LEFT	0xcf
#define LS_SYM_RIGHT	0xc8
#define LS_SYM_DOT	0xdc
#define LS_SYM_VOLTS	0xc6
// lcd segment addresses for menu
#define LS_MENU_MODEL	0xd0
#define LS_MENU_NAME	0xcd
#define LS_MENU_REV	0xcb
#define LS_MENU_EPO	0xca
#define LS_MENU_TRIM	0xd1
#define LS_MENU_DR	0xce
#define LS_MENU_EXP	0xcc
#define LS_MENU_ABS	0xc9




// manipulating group of segment together by applying bitmaps
extern void lcd_set(u8 id, u8 *bitmap);
extern void lcd_set_blink(u8 id, u8 on_off);

// IDs for lcd_set(), lcd_set_blink()
#define LCHR1		0
#define LCHR2		1
#define LCHR3		2
#define L7SEG		3
#define LMENU		4
// special (fake) bitmaps for lcd_set()
#define LB_EMPTY	(u8 *)0xff00
#define LB_FULL		(u8 *)0xffff




// high level functions to write chracters and numbers
#define LCHAR_MINUS_ONE ('Z' + 1)
#define LCHAR_ONE_ZERO  ('Z' + 2)
#define LCHAR_MIN ' '
#define LCHAR_MAX LCHAR_ONE_ZERO
#define L7SEG_MAX 15
extern void lcd_char(u8 id, u8 c);  // id the same as in lcd_set()
extern void lcd_chars(u8 *chars);   // 3 chars
extern void lcd_chars2(u8 *chars);  // 2 chars to chars 2+3
extern void lcd_char_num3(s16 num); // num -199...1099 to 3 chars
extern void lcd_char_num2_lbl(s8 num, u8 *labels);  // num -99..99 with labels for <0, =0, >0
extern void lcd_7seg(u8 number);    // num 0-15 (>=10 as hexa numbers)
extern void lcd_menu(u8 menus);     // OR-ed selected menus
extern void lcd_chars_blink(u8 on_off);  // lcd_set_blink() for 3 chars
extern void lcd_chars_blink_mask(u8 on_off, u8 mask);  // lcd_set_blink() for 3 chars masked by bits
#define LB_CHR1 	0b00000001
#define LB_CHR2 	0b00000010
#define LB_CHR3 	0b00000100

// values for lcd_menu()
#define LM_MODEL	0x80
#define LM_NAME		0x40
#define LM_REV		0x20
#define LM_EPO		0x10
#define LM_TRIM		0x08
#define LM_DR		0x04
#define LM_EXP		0x02
#define LM_ABS		0x01

// values for lcd_7seg() - IDs of 7seg letters
#define L7_A		10
#define L7_B		11
#define L7_C		12
#define L7_D		13
#define L7_E		14
#define L7_F		15
#define L7_G		16
#define L7_H		17
#define L7_I		18
#define L7_J		19
#define L7_K		20
#define L7_L		21
#define L7_M		22
#define L7_N		23
#define L7_O		24
#define L7_P		25
#define L7_R		26
#define L7_T		27
#define L7_U		28
#define L7_V		29
#define L7_Y		30
#define L7_MINUS	31




// update/clear/set
extern void lcd_update(void);  // this one actually write to display
#define lcd_update_pause()  lcd_update(); pause()
#define lcd_update_stop()   lcd_update(); stop()
extern void lcd_clear(void);
extern void lcd_set_full_on(void);




// for lcd_blink_cnt, maximal value and value when changing to space
// in 5ms steps
#define LCD_BLNK_CNT_MAX	160
#define LCD_BLNK_CNT_BLANK	100
extern volatile _Bool lcd_blink_flag;
extern volatile u8    lcd_blink_cnt;
extern volatile _Bool lcd_blink_something;




// for backlight
#define LCD_BCK0 BRES(PD_ODR, 2)
#define LCD_BCK1 BSET(PD_ODR, 2)
extern _Bool lcd_bck_on;
extern u16   lcd_bck_count;

#define BACKLIGHT_MAX  (u16)(0xffff)
extern void backlight_set_default(u16 seconds);
extern void backlight_on_sec(u16 seconds);
extern void backlight_on(void);
extern void backlight_off(void);




// LCD task
E_TASK(LCD);


#endif

